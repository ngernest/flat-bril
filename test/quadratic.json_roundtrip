{
  "functions": [
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        },
        {
          "name": "c",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a",
            "b",
            "c"
          ],
          "funcs": [
            "quadratic"
          ],
          "labels": [],
          "op": "call"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "v1",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v1"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.cond.0"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v4",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v3",
            "v4"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v2",
            "v5"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "v6"
          ],
          "funcs": [],
          "labels": [
            "for.body.0",
            "for.end.0"
          ],
          "op": "br"
        },
        {
          "label": "for.body.0"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v9",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v8",
            "v9"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v11",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v10",
            "v11"
          ],
          "dest": "v12",
          "funcs": [],
          "labels": [],
          "op": "ge",
          "type": "bool"
        },
        {
          "args": [
            "v12"
          ],
          "funcs": [],
          "labels": [
            "then.7",
            "else.7"
          ],
          "op": "br"
        },
        {
          "label": "then.7"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v13"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "else.7"
        },
        {
          "label": "endif.7"
        },
        {
          "args": [
            "i"
          ],
          "dest": "v14",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v15",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v14",
            "v15"
          ],
          "dest": "v16",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v16"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.cond.0"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end.0"
        },
        {
          "dest": "v17",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v17"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "sqrt",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        },
        {
          "name": "c",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "b"
          ],
          "dest": "v0",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "b"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v0",
            "v1"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "dest": "v3",
          "op": "const",
          "type": "int",
          "value": 4
        },
        {
          "args": [
            "a"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v3",
            "v4"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "c"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v5",
            "v6"
          ],
          "dest": "v7",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v2",
            "v7"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v8"
          ],
          "dest": "s",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v9",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "a"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v9",
            "v10"
          ],
          "dest": "v11",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v11"
          ],
          "dest": "d",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v12",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "b"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v12",
            "v13"
          ],
          "dest": "v14",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "s"
          ],
          "dest": "v15",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v15"
          ],
          "dest": "v16",
          "funcs": [
            "sqrt"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v14",
            "v16"
          ],
          "dest": "v17",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v17"
          ],
          "dest": "r1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v18",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "b"
          ],
          "dest": "v19",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v18",
            "v19"
          ],
          "dest": "v20",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "s"
          ],
          "dest": "v21",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v21"
          ],
          "dest": "v22",
          "funcs": [
            "sqrt"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v20",
            "v22"
          ],
          "dest": "v23",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v23"
          ],
          "dest": "r2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "r1"
          ],
          "dest": "v24",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "d"
          ],
          "dest": "v25",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v24",
            "v25"
          ],
          "dest": "v26",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v26"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v27",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "r2"
          ],
          "dest": "v28",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "d"
          ],
          "dest": "v29",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v28",
            "v29"
          ],
          "dest": "v30",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v30"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v31",
          "op": "const",
          "type": "int",
          "value": 0
        }
      ],
      "name": "quadratic"
    }
  ]
}
