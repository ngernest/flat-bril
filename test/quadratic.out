@main
	op: call 		args: ["a", "b", "c"]		funcs: ["quadratic"]
@sqrt
	op: const		dest: v1		type: int		value: 1	
	op: id   		dest: i 		type: int		args: ["v1"]	
.for.cond.0
	op: id   		dest: v2		type: int		args: ["i"]	
	op: id   		dest: v3		type: int		args: ["x"]	
	op: const		dest: v4		type: int		value: 1	
	op: sub  		dest: v5		type: int		args: ["v3", "v4"]	
	op: lt   		dest: v6		type: bool		args: ["v2", "v5"]	
	op: br   		args: ["v6"]		labels: ["for.body.0", "for.end.0"]
.for.body.0
	op: id   		dest: v8		type: int		args: ["i"]	
	op: id   		dest: v9		type: int		args: ["i"]	
	op: mul  		dest: v10		type: int		args: ["v8", "v9"]	
	op: id   		dest: v11		type: int		args: ["x"]	
	op: ge   		dest: v12		type: bool		args: ["v10", "v11"]	
	op: br   		args: ["v12"]		labels: ["then.7", "else.7"]
.then.7
	op: id   		dest: v13		type: int		args: ["i"]	
	op: ret  		args: ["v13"]	
.else.7
.endif.7
	op: id   		dest: v14		type: int		args: ["i"]	
	op: const		dest: v15		type: int		value: 1	
	op: add  		dest: v16		type: int		args: ["v14", "v15"]	
	op: id   		dest: i 		type: int		args: ["v16"]	
	op: jmp  		labels: ["for.cond.0"]
.for.end.0
	op: const		dest: v17		type: int		value: 0	
	op: ret  		args: ["v17"]	
@quadratic
	op: id   		dest: v0		type: int		args: ["b"]	
	op: id   		dest: v1		type: int		args: ["b"]	
	op: mul  		dest: v2		type: int		args: ["v0", "v1"]	
	op: const		dest: v3		type: int		value: 4	
	op: id   		dest: v4		type: int		args: ["a"]	
	op: mul  		dest: v5		type: int		args: ["v3", "v4"]	
	op: id   		dest: v6		type: int		args: ["c"]	
	op: mul  		dest: v7		type: int		args: ["v5", "v6"]	
	op: sub  		dest: v8		type: int		args: ["v2", "v7"]	
	op: id   		dest: s 		type: int		args: ["v8"]	
	op: const		dest: v9		type: int		value: 2	
	op: id   		dest: v10		type: int		args: ["a"]	
	op: mul  		dest: v11		type: int		args: ["v9", "v10"]	
	op: id   		dest: d 		type: int		args: ["v11"]	
	op: const		dest: v12		type: int		value: 0	
	op: id   		dest: v13		type: int		args: ["b"]	
	op: sub  		dest: v14		type: int		args: ["v12", "v13"]	
	op: id   		dest: v15		type: int		args: ["s"]	
	op: call 		dest: v16		type: int		args: ["v15"]		funcs: ["sqrt"]
	op: add  		dest: v17		type: int		args: ["v14", "v16"]	
	op: id   		dest: r1		type: int		args: ["v17"]	
	op: const		dest: v18		type: int		value: 0	
	op: id   		dest: v19		type: int		args: ["b"]	
	op: sub  		dest: v20		type: int		args: ["v18", "v19"]	
	op: id   		dest: v21		type: int		args: ["s"]	
	op: call 		dest: v22		type: int		args: ["v21"]		funcs: ["sqrt"]
	op: sub  		dest: v23		type: int		args: ["v20", "v22"]	
	op: id   		dest: r2		type: int		args: ["v23"]	
	op: id   		dest: v24		type: int		args: ["r1"]	
	op: id   		dest: v25		type: int		args: ["d"]	
	op: div  		dest: v26		type: int		args: ["v24", "v25"]	
	op: print		args: ["v26"]	
	op: const		dest: v27		type: int		value: 0	
	op: id   		dest: v28		type: int		args: ["r2"]	
	op: id   		dest: v29		type: int		args: ["d"]	
	op: div  		dest: v30		type: int		args: ["v28", "v29"]	
	op: print		args: ["v30"]	
	op: const		dest: v31		type: int		value: 0	
